flatgfa/src/packedseq.rs

```diff
@@ -0,0 +1,328 @@
+    }
+}
+
+struct PackedVecIterator<'a> {
+    data: &'a Vec<u8>,
+    cur_index: usize,
+    cur_high_nibble: bool,
+    high_nibble_end: bool,
+    last: bool,
+}
+
+impl<'a> PackedVecIterator<'a> {
+    pub fn new(vec: &'a PackedVec) -> Self {
+        Self {
+            data: &vec.data,
+            cur_index: 0,
+            cur_high_nibble: false,
+            high_nibble_end: vec.high_nibble_end,
+            last: false,
+        }
+    }
+}
+
+impl<'a> Iterator for PackedVecIterator<'a> {
+    type Item = u8;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.cur_index >= self.data.len() || self.last {
+            return None;
+        }
+        if self.cur_index == self.data.len() - 1 && !self.high_nibble_end {
+            self.last = true;
+        }
+        let result = if self.cur_high_nibble {
+            (self.data[self.cur_index] >> 4) & 0b00001111u8
+        } else {
+            self.data[self.cur_index] & 0b00001111u8
+        };
+        if self.cur_high_nibble {
+            self.cur_index += 1;
+        }
+        self.cur_high_nibble = !self.cur_high_nibble;
+        Some(result)
+    }
+}
+
+/// A reference to a subsection of a nucleotide sequence stored in a PackedVec
+/// data_ref: The underlying vector that stores the sequence referenced by this slice
+/// span: The specific section of the sequence that this slice references
+struct PackedSlice<'a> {
+    vec_ref: &'a PackedVec,
+    span: std::ops::Range<u8>,
+}
+
+fn get_seq(vec_ref: &PackedVec, span: std::ops::Range<u8>) -> Vec<Nucleotide> {
+    let mut arr: Vec<Nucleotide> = Vec::with_capacity((span.end - span.start).into());
+    for i in span.start..=span.end {
+        arr.push(get_vec_elem(vec_ref, i));
+    }
+    return arr;
+}
+
+/// Returns the element of *seq* at index *index*
+fn get_vec_elem(seq: &PackedVec, index: u8) -> Nucleotide {
+    if index % 2 == 1 {
+        let i: usize = (index / 2) as usize;
+        return ((&seq.data[i] & 0b11110000u8) >> 4).into();
+    } else {
+        let i: usize = (index / 2) as usize;
+        return (&seq.data[i] & 0b00001111u8).into();
+    }
+}
+
+/// Sets the element of *seq* at index *index* to *elem*
+fn set_vec_elem(seq: &mut PackedVec, index: u8, input: Nucleotide) {
+    let elem: u8 = input.into();
+    if index % 2 == 1 {
+        let i: usize = (index / 2) as usize;
+        println!("i: {}", i);
+        seq.data[i] = (0b00001111u8 & seq.data[i]) | (elem << 4);
+    } else {
+        let i: usize = (index / 2) as usize;
+        seq.data[i] = (0b11110000u8 & seq.data[i]) | elem;
+    }
+}
+
+/// Returns a uncompressed vector that contains the sequence in *seq*
+fn get_vec_seq(seq: &PackedVec) -> Vec<Nucleotide> {
+    let end_index = if (seq.high_nibble_end) {
+        (seq.data.len() * 2) - 1
+    } else {
+        (seq.data.len() * 2) - 2
+    };
+    let span: std::ops::Range<u8> = std::ops::Range {
+        start: 0,
+        end: end_index as u8,
+    };
```

Comment:
Instead of using an explicit struct constructor here, Rust has special syntax for making `Range` objects! The expression `x..y` is equivalent to `Range { start: x, end: y }`. So you can probably just make this `0..self.len()`.